flow_post_lda.scala

```diff
@@ -1,270 +1,543 @@
+Logger.getLogger("akka").setLevel(Level.OFF)
+
+def isNumeric(input: String): Boolean = {
+    if (input == "") {false}
+    else {input.forall(x => (x.isDigit || x == '.' || x =='E' || x == 'e')) && input.count(_ == '.')<2 && (input.count(_ == 'E') + input.count(_ == 'e')) <2}
+}
+def toDouble(s: String, default:Double=Double.NaN) = {if (isNumeric(s)) s.toDouble else default}
+//def isHeader(line: Array[String], header: Array[String]) = line.deep == header.deep
+
+def removeHeader(input: org.apache.spark.rdd.RDD[String]) = {
+    val header = input.first
+    val output = input.filter(line => !(line == header))
+    output
+}
+
+def matrixToRdd(m: org.apache.spark.mllib.linalg.Matrix) : org.apache.spark.rdd.RDD[(Long, org.apache.spark.mllib.linalg.DenseVector)] = {
+    val cols = m.toArray.grouped(m.numRows)
+    val rows = cols.toSeq.transpose
+    val vectors = rows.zipWithIndex.map{case (row, index) => (index.toLong, new org.apache.spark.mllib.linalg.DenseVector(row.toArray))}
+    sc.parallelize(vectors)
+}
+
+
+def print(input: org.apache.spark.rdd.RDD[String]) = input.take(10).foreach(println)
+
+
+// Load and parse the data
+
+class SimpleCSVHeader(header:Array[String]) extends Serializable {
+  val index = header.zipWithIndex.toMap
+  def apply(array:Array[String], key:String):String = array(index(key))
+}
+
+
+
+
+
+//----------Inputs-------------
+val file = System.getenv("DPATH")
+val topic_mix_file = System.getenv("HPATH") + "/doc_results.csv"
+val pword_file = System.getenv("HPATH")+"/word_results.csv"
+val scored_output_file = System.getenv("HPATH") + "/scored"
+val threshold : Double = System.getenv("TOL").toDouble
+
+val compute_quantiles : Boolean = true
+val quant = Array(0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9)
+val quint = Array(0, 0.2, 0.4, 0.6, 0.8)
+var ibyt_cuts = new Array[Double](10)
+var ipkt_cuts = new Array[Double](5)
+var time_cuts = new Array[Double](10)
+
+//-----------------------------
+
+def compute_ecdf(x : org.apache.spark.rdd.RDD[Double]) : org.apache.spark.rdd.RDD[(Double, Double)] ={
+    val counts = x.map( v => (v,1)).reduceByKey(_+_).sortByKey().cache()
+    // compute the partition sums
+    val partSums: Array[Double] = 0.0 +: counts.mapPartitionsWithIndex {
+      case (index, partition) => Iterator(partition.map { case (sample, count) => count }.sum.toDouble)
+    }.collect()
+    
+    // get sample size
+    val numValues = partSums.sum
+    
+    // compute empirical cumulative distribution
+    val sumsRdd = counts.mapPartitionsWithIndex {
+      case (index, partition) => {
+        var startValue = 0.0
+        for (i <- 0 to index) {
+          startValue += partSums(i)
+        }
+        partition.scanLeft((0.0, startValue))((prev, curr) => (curr._1, prev._2 + curr._2)).drop(1)
+      }
+    }
+    sumsRdd.map( elem => (elem._1, elem._2 / numValues))
+}
+
+def distributed_quantiles(quantiles: Array[Double], ecdf: org.apache.spark.rdd.RDD[(Double, Double)]): Array[Double] ={
```

Comment:
spark 1.6 has built-in functionality like this,  
in future we should look to moving to their implementations (put debug, maintenance and performance on them)

https://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.ml.feature.QuantileDiscretizer

(it is experimental in 1.6.1 so take w/ grain of salt)